### ISSUE TITLE
Bug: BAM->SAM without NUL-terminated header

### ISSUE THREAD

**Issue by nathanhaigh** [samtools/samtools#661]
An alignment tool I use [BioKanga](https://github.com/csiro-crop-informatics/biokanga) does not NUL-terminate the header in its BAM output files. According to the spec, it is valid that the header does not need to be NUL-terminated: https://github.com/samtools/hts-specs/blob/master/SAMv1.tex#L795.

When I do a BAM->SAM conversion of such a file, samtools does not deal with this correctly. Samtools fails to correctly place the first alignment on the next line following the header and instead places it at the end of the last header line. For example:

```
@PG     ID:biokanga     VN:4.3.5HWI-ST226:209:D04H2ACXX:4:1107:8119:199917      99      chr1A   293     255     62M     =       484     269     CTCGAGCT
```

**Comment by ste69r** [comment]
To clarify this samtools bug report, the issue is not as to if the header section is terminated with the ASCII NUL '\0' character (optional) but rather that samtools is incorrectly handling the absence of a new line character ’\n’ immediately preceding any NUL at the end of the header block when converting BAM back to SAM format. Samtools is blindly prepending the BAM header section to the transformed first alignment when converting to SAM. Thus, if no new line character terminating the header section then the observed concatenation of the last line of header with the first alignment.

**Comment by jmarshall** [comment]
The spec is not explicit on this, just referring to “plain header text in SAM”; however the header text is made out of lines, and (on Unix and in C) lines are considered to be newline-**terminated** (including the last one).

From preliminary testing, it appears that (possibly by luck) HTSJDK accepts such BAM files.  What do other implementations do?  Before accepting this as a samtools/htslib bug, there is a decision to be made as to whether such BAM files are valid.  Either way, a small spec clarification is warranted.  (And what about CRAM files?)

**Comment by jkbonfield** [comment]
The only reasonable choices here are whether we should fix the missing newline or error, but continuing regardless and producing invalid output is incorrect.  Given the weakness of the spec I'm in favour of automatically fixing.  I'll check CRAM.

**Comment by jmarshall** [comment]
Yes, but deciding between accepting-without-final-newline and error is properly an interoperability question.  Does starting to accept such files now cause problems for other implementations?  That is the way to decide on a course of action, not some subjective “spec weakness”.

As a byproduct of its super-factored implementation :smile:, HTSJDK does indeed accept this in all cases (see _StringLineReader.java_).  This probably mostly answers the interoperability question, but we may care about the minor implementations…

**Comment by jkbonfield** [comment]
At the very least I was considering a warning: Eg "Missing trailing newline in header - truncated?", but am happy to make it a hard error too.  I'll explore via samtools reheader.

**Comment by ste69r** [comment]
a) All text lines are not considered, by definition, to be NL terminated on Unix and in C as stated by jmarshall. Individual text lines are normally expected to be NUL terminated. Multiple text lines which are part of a composite text body may use NL as line separators but there is no expectation that the last line in that composite text body must contain a NL immediately preceding the terminating NUL.
b) Processing applications should be robust and be prepared to accept, in the absence of an explicit format specification statement (in this instance the SAM specification), the presence/absence of a header terminating NL character; after all the spec explicitly states that one or more NUL terminators are optional and there is a header body length field which is to include the number of optional NULs as part of the body length. This I hope is handled correctly and NULs are not present in the BAM to SAM conversion. 

The correct solution is to check for the existence of a NL terminating the body text and truncate the body text at the 1st terminating NL instance (nothing in the spec to prevent multiple terminating NLs) or at the first NUL (multiples of these are allowed), which ever occurs first in the header body. Then precede the 1st alignment record with a NL when outputting as SAM. This is a robust solution and will not break existing downstream applications.

Yes, the format specification should be tightened, but applications should also be robust, with few assumptions made by processing applications. I would actually argue that the header body should be always  parsed and checked for conformance, downstream applications may have dependencies on the content of the header. For example is there a check for consistency in sequence names/lengths in the BAM sequence section relative to those in the header ???? .....

**Comment by jmarshall** [comment]
@ste69r: (a) That's just not true.  [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206) defines a _line_ as “A sequence of zero or more non-\<newline\> characters plus a terminating \<newline\> character” and a _text file_ as a file that contains characters organized into zero or more lines.  The C standard §5.1.1.2 (Translation phases) specifies that “A source file that is not empty shall end in a new-line character, which shall not be immediately preceded by a backslash character […]”.

That many modern implementations of C compilers and various text tools (like sed, awk, etc) also accept files whose last line does not end in a newline is a Quality Of Implementation extension for user convenience.  I remember implementations of such tools that did not, and I expect they still exist on user-hostile systems like Solaris.

That said, it's probably appropriate to be similarly generous in BAM headers (though this is a text blob within a binary file, so the user-editing convenience arguments do not apply), but **that choice should be solely guided by interoperability concerns**.

Or you lot could just fix this BioKanga newline bug!

**Comment by ste69r** [comment]
I have no wish to become entangled in an argument about language semantics.
The header section is not a file, there is no stated requirement in the SAM specification for it to be NL terminated; you stated  that the header is a ‘”text blob within a binary file”. Termination of a text blob is arbitrary and this statement is supported by the presence of an explicit header length specifier.

So, can you please confirm as to how Samtools currently handles BAM header section terminator(s), NLs or NULs and possible multiples of same. Do you parse the header ‘blob’ and validate it’s contents when converting BAM to SAM. Do you simply prepend the header ‘blob’ to the first alignment?

Yes, it would be trivial to ensure a NL is appended to the header in Biokanga when generating BAM, and that would ensure interoperability with Samtools when Samtools is converting to SAM, but what about other applications attempting to convert BAM to SAM?

Finally, this is not a bug in BioKanga (it is adhering to the specification) and I object to your use of the term ‘you lot’. Please, let’s keep this discussion professional.


Marshall [mailto:notifications@github.com]
Sent: Monday, April 10, 2017 8:39 PM
To: samtools/samtools <samtools@noreply.github.com>
Cc: Stuart Stephen <stuartjs@g3web.com>; Mention <mention@noreply.github.com>
Subject: Re: [samtools/samtools] Bug: BAM->SAM without NUL-terminated header (#661)


@ste69r<https://github.com/ste69r>: (a) That's just not true. POSIX<http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206> defines a line as “A sequence of zero or more non-<newline> characters plus a terminating <newline> character” and a text file as a file that contains characters organized into zero or more lines. The C standard §5.1.1.2 (Translation phases) specifies that “A source file that is not empty shall end in a new-line character, which shall not be immediately preceded by a backslash character […]”.

That many modern implementations of C compilers and various text tools (like sed, awk, etc) also accept files whose last line does not end in a newline is a Quality Of Implementation extension for user convenience. I remember implementations of such tools that did not, and I expect they still exist on user-hostile systems like Solaris.

That said, it's probably appropriate to be similarly generous in BAM headers (though this is a text blob within a binary file, so the user-editing convenience arguments do not apply), but that choice should be solely guided by interoperability concerns.

Or you lot could just fix this BioKanga newline bug!

**Comment by jkbonfield** [comment]
There is no prepending of header to the first record.  Converting BAM to SAM is simply a case of write header followed by write each record in turn.  The "error" here is that there isn't a check for header validity; either multiple newlines, lines starting without @, or a lack of trailing newline.

My inclination here is to fix these on reading so the in-memory data structure is correct. Whether or not to emit a warning is debatable.

For what it's worth, I tested picard ViewSam and format conversion, Biobambam2's bamcat, and Scramble to view a BAM file with missing trailing newline and all of them silently fix this.  I suspect it's because all of them parse and perform minimal validation of the header.  Htslib just loads it as a blob of text and only parses it if and when required, so BAM to SAM transcodes any BAM formatting issues to SAM formatting issues.

**Comment by ste69r** [comment]
Presumably this fix (Trivial SAM header sanitising #509) also applies to samtools quickcheck -vvv ${BAM_FILE} which is reporting no errors when validating BAM headers as generated by BioKanga?

**Comment by jmarshall** [comment]
I have now also examined sambamba/BioD, biogo-hts, and bamtools.  Like HTSJDK, BioD and biogo read via essentially `foreach (split /\n/) { continue if length < 3; … }` and bamtools reads via `std::getline`, so all four accept either with or without a final newline.  All four of them always write a final newline (though a bamtools API user might conceivably change the header text string before writing it out).

I see BioKanga has now fixed its bug and always writes a final newline like these other implementations.

**Comment by jkbonfield** [comment]
@jmarshall Fixing the issue in both places sounds like the robust thing to do, so I'm happy with this outcome.

@ste69r The quickcheck tool uses ```sam_hdr_read``` at verbosity -vvvv or above, so it then runs the sanitise function.  Without it the checks are whatever it does internally, which doesn't appear to include newline checking.
